runtime/compiler/aarch64/codegen/J9ARM64Snippet.cpp

```diff
@@ -23,88 +23,218 @@
    TR::LabelSymbol *callLabel,
    TR::LabelSymbol *restartLabel)
    : _incLabel(incLabel),
-     _lwOffset(lwOffset),
      TR::ARM64HelperCallSnippet(codeGen, monitorNode, callLabel, monitorNode->getSymbolReference(), restartLabel)
    {
-   TR_UNIMPLEMENTED();
+   // Helper call, preserves all registers
+   incLabel->setSnippet(this);
+   gcMap().setGCRegisterMask(0xFFFFFFFF);
    }
 
 uint8_t *
 TR::ARM64MonitorEnterSnippet::emitSnippetBody()
    {
-   TR_UNIMPLEMENTED();
+   // The AArch64 code for the snippet looks like:
+   //
+   // incLabel:
+   //    and     tempReg, dataReg, ~(OBJECT_HEADER_LOCK_BITS_MASK - OBJECT_HEADER_LOCK_LAST_RECURSION_BIT)
+   //    cmp     metaReg, tempReg
+   //    bne     callLabel
+   //    add     dataReg, dataReg, LOCK_INC_DEC_VALUE
+   //    str     dataReg, [addrReg]
+   //    b       restartLabel
+   // callLabel:
+   //    bl      jitMonitorEntry
+   //    b       restartLabel
+
+   TR::RegisterDependencyConditions *deps = getRestartLabel()->getInstruction()->getDependencyConditions();
+
+   TR::RealRegister *metaReg = cg()->getMethodMetaDataRegister();
+   TR::RealRegister *dataReg = cg()->machine()->getRealRegister(deps->getPostConditions()->getRegisterDependency(1)->getRealRegister());
+   TR::RealRegister *addrReg = cg()->machine()->getRealRegister(deps->getPostConditions()->getRegisterDependency(2)->getRealRegister());
+   TR::RealRegister *tempReg = cg()->machine()->getRealRegister(deps->getPostConditions()->getRegisterDependency(3)->getRealRegister());
+   TR::RealRegister *zeroReg = cg()->machine()->getRealRegister(TR::RealRegister::xzr);
+   TR::InstOpCode::Mnemonic op;
 
    uint8_t *buffer = cg()->getBinaryBufferCursor();
+
+   _incLabel->setCodeLocation(buffer);
+
+   *(int32_t *)buffer = TR::InstOpCode::getOpCodeBinaryEncoding(TR::InstOpCode::andimmx);
+   tempReg->setRegisterFieldRD((uint32_t *)buffer);
+   dataReg->setRegisterFieldRN((uint32_t *)buffer);
+   *(int32_t *)buffer |= 0x196000; // immr=25, imms=24 for 0xFFFFFF80
+   buffer += ARM64_INSTRUCTION_LENGTH;
+
+   *(int32_t *)buffer = TR::InstOpCode::getOpCodeBinaryEncoding(TR::InstOpCode::subsx); // for cmp
+   metaReg->setRegisterFieldRN((uint32_t *)buffer);
+   tempReg->setRegisterFieldRM((uint32_t *)buffer);
+   zeroReg->setRegisterFieldRD((uint32_t *)buffer);
+   buffer += ARM64_INSTRUCTION_LENGTH;
+
+   *(int32_t *)buffer = TR::InstOpCode::getOpCodeBinaryEncoding(TR::InstOpCode::b_cond);
+   *(int32_t *)buffer |= ((4 << 5) | TR::CC_NE); // 4 instructions forward, bne
+   buffer += ARM64_INSTRUCTION_LENGTH;
+
+   *(int32_t *)buffer = TR::InstOpCode::getOpCodeBinaryEncoding(TR::InstOpCode::addimmx);
+   dataReg->setRegisterFieldRD((uint32_t *)buffer);
+   dataReg->setRegisterFieldRN((uint32_t *)buffer);
+   *(int32_t *)buffer |= ((LOCK_INC_DEC_VALUE & 0xFFF) << 10); // imm12
+   buffer += ARM64_INSTRUCTION_LENGTH;
+
+   op = fej9->generateCompressedLockWord() ? TR::InstOpCode::strimmw : TR::InstOpCode::strimmx;
```

Comment:
compilation fails because `fej9` is not declared in this function.

Reply:
Thank you for pointing it out.  I had put a wrong version before fix into the commit.